### PR TITLE
Do step 15 - Add tests for a ShoppingCart functionality. Level Servic…

### PR DESCRIPTION
**Pull request by yurlis**
…e and Repository - ok. I tried doing with Mock Service and Repository levels, but nothing worked. Unfortunately.

### CODE REVIEW

**Comment by yurlis** [comment]
shopping-cart-coverage
Kind of code coverage is shown for the Shopping Cart, but I don't know why it doesn't show for the repository tests.

File: src/main/java/com/bookstoreapp/service/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -23,13 +28,27 @@ public class ShoppingCartServiceImpl implements ShoppingCartService {
...
         ShoppingCart cartFromDB = getCartFromDB(userId);
-        CartItem requestedItemToAdd = itemMapper.toModel(requestDto);
-        requestedItemToAdd.setShoppingCart(cartFromDB);
-        itemRepository.save(requestedItemToAdd);
+
+        Optional<CartItem> cartItem = itemRepository
+                .findByShoppingCartIdAndBookId(cartFromDB.getId(), requestDto.bookId());
+
+        if (cartItem.isPresent()) {
+            BookAlreadyExistsException.supplier("Book already exists in the cart");
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstoreapp/service/ShoppingCartServiceImpl.java]: it's better to increase quantity of book in this case
what if user would like to add 2 books? ))

File: src/main/java/com/bookstoreapp/repository/role/RoleRepository.java [hunk 2]
```diff
@@ -2,10 +2,15 @@
...
+import org.springframework.data.repository.query.Param;
 import org.springframework.stereotype.Repository;
 import java.util.Optional;
 
 @Repository
 public interface RoleRepository extends JpaRepository<Role, Long> {
     Optional<Role> findRoleByRole(Role.RoleType roleName);
+
+    @Query("FROM Role r WHERE r.role = :role_name ")
+    Role getRoleByRoleType(@Param("role_name") Role.RoleType roleType);
```
Comment (Elena-Bruyako) [inline src/main/java/com/bookstoreapp/repository/role/RoleRepository.java]: Why do you need this method?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Why there are several versions of ShopCartSertvice tests? 
When you test controller layer, you need to use integration tests only
Let's discuss in chat your implementation

**Comment by yurlis** [comment]
full-application-coverage
Test coverage of controller and service level classes

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
I answered your questions regarding Olena's comments in the discussion on the platform. So first let's come to a common decision and then make a request for a recheck.
I also see that in the new commit you added the findByEmail method to the service leyer and tested it, but I didn't find any use for this method, why is it needed?

File: src/test/java/com/bookstoreapp/controller/ShoppingCartControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+import static org.mockito.Mockito.when;
+import static org.springframework.security.test.web.servlet.setup.SecurityMockMvcConfigurers.springSecurity;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
+import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@AutoConfigureMockMvc
+public class ShoppingCartControllerTest {
```
Comment (Elena-Bruyako) [inline src/test/java/com/bookstoreapp/controller/ShoppingCartControllerTest.java]: unit test for controller still here, please remove it

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]

**Pull request merged**